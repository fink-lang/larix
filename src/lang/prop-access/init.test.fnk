{describe, it, expect, to_match_snapshot} = import '@fink/jest/test.fnk'

{parse, parse_expr} = import '../../parser.fnk'


describe 'prop access', fn:

  it 'parses single line', fn:
    expect
      parse_expr 'foo.bar'
      to_match_snapshot


  it 'parses various idents', fn:
    expect
      parse_expr 'foo.π'
      to_match_snapshot

    expect
      parse_expr 'foo.$spam'
      to_match_snapshot

    expect
      parse_expr 'foo._spam'
      to_match_snapshot

    expect
      parse_expr 'foo.∑'
      to_match_snapshot


  it 'parses with taken keywords', fn:
    expect
      parse_expr 'foo.match'
      to_match_snapshot

    expect
      parse_expr 'foo.fn'
      to_match_snapshot


  it 'parses member with str expr', fn:
    expect
      parse_expr "foo.'bar spam'"
      to_match_snapshot


  it 'parses member with group expr', fn:
    expect
      parse_expr 'foo.(key)'
      to_match_snapshot


  it 'errors with invalid member', fn:
    expect
      parse 'foo.[bar]', 'test.fnk'
      to_match_snapshot

    expect
      parse 'foo.+bar', 'test.fnk'
      to_match_snapshot

    expect
      parse 'foo.-bar', 'test.fnk'
      to_match_snapshot

    expect
      parse 'foo.*bar', 'test.fnk'
      to_match_snapshot

    expect
      parse 'foo./bar', 'test.fnk'
      to_match_snapshot

    expect
      parse 'foo.?bar', 'test.fnk'
      to_match_snapshot

    expect
      parse 'foo.:bar', 'test.fnk'
      to_match_snapshot

    expect
      parse 'foo.{bar', 'test.fnk'
      to_match_snapshot

    expect
      parse 'foo.}bar', 'test.fnk'
      to_match_snapshot

    expect
      parse 'foo.[bar', 'test.fnk'
      to_match_snapshot

    expect
      parse 'foo.]bar', 'test.fnk'
      to_match_snapshot

    expect
      parse 'foo.)bar', 'test.fnk'
      to_match_snapshot

    expect
      parse 'foo.,bar', 'test.fnk'
      to_match_snapshot



