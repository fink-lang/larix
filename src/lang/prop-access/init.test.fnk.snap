// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prop access errors with invalid member 1`] = `
module (1:0-1:4)
  ident (1:0-1:3) foo

error:
test.fnk:1:4
1| foo.[bar]
       ^

Expected member key to be an ident, string or group expr.



`;

exports[`prop access errors with invalid member 2`] = `
module (1:0-1:4)
  ident (1:0-1:3) foo

error:
test.fnk:1:4
1| foo.+bar
       ^

Expected member key to be an ident, string or group expr.



`;

exports[`prop access errors with invalid member 3`] = `
module (1:0-1:4)
  ident (1:0-1:3) foo

error:
test.fnk:1:4
1| foo.-bar
       ^

Expected member key to be an ident, string or group expr.



`;

exports[`prop access errors with invalid member 4`] = `
module (1:0-1:4)
  ident (1:0-1:3) foo

error:
test.fnk:1:4
1| foo.*bar
       ^

Expected member key to be an ident, string or group expr.



`;

exports[`prop access errors with invalid member 5`] = `
module (1:0-1:4)
  ident (1:0-1:3) foo

error:
test.fnk:1:4
1| foo./bar
       ^

Expected member key to be an ident, string or group expr.



`;

exports[`prop access errors with invalid member 6`] = `
module (1:0-1:4)
  ident (1:0-1:3) foo

error:
test.fnk:1:4
1| foo.?bar
       ^

Expected member key to be an ident, string or group expr.



`;

exports[`prop access errors with invalid member 7`] = `
module (1:0-1:4)
  ident (1:0-1:3) foo

error:
test.fnk:1:4
1| foo.:bar
       ^

Expected member key to be an ident, string or group expr.



`;

exports[`prop access errors with invalid member 8`] = `
module (1:0-1:4)
  ident (1:0-1:3) foo

error:
test.fnk:1:4
1| foo.{bar
       ^

Expected member key to be an ident, string or group expr.



`;

exports[`prop access errors with invalid member 9`] = `
module (1:0-1:4)
  ident (1:0-1:3) foo

error:
test.fnk:1:4
1| foo.}bar
       ^

Expected member key to be an ident, string or group expr.



`;

exports[`prop access errors with invalid member 10`] = `
module (1:0-1:4)
  ident (1:0-1:3) foo

error:
test.fnk:1:4
1| foo.[bar
       ^

Expected member key to be an ident, string or group expr.



`;

exports[`prop access errors with invalid member 11`] = `
module (1:0-1:4)
  ident (1:0-1:3) foo

error:
test.fnk:1:4
1| foo.]bar
       ^

Expected member key to be an ident, string or group expr.



`;

exports[`prop access errors with invalid member 12`] = `
module (1:0-1:4)
  ident (1:0-1:3) foo

error:
test.fnk:1:4
1| foo.)bar
       ^

Expected member key to be an ident, string or group expr.



`;

exports[`prop access errors with invalid member 13`] = `
module (1:0-1:4)
  ident (1:0-1:3) foo

error:
test.fnk:1:4
1| foo.,bar
       ^

Expected member key to be an ident, string or group expr.



`;

exports[`prop access parses member with group expr 1`] = `
member . (1:0-1:9)
  ident (1:0-1:3) foo
  group (1:4-1:9)
    ident (1:5-1:8) key
`;

exports[`prop access parses member with str expr 1`] = `
member . (1:0-1:14)
  ident (1:0-1:3) foo
  string ' (1:4-1:14)
    'bar spam'
`;

exports[`prop access parses single line 1`] = `
member . (1:0-1:7)
  ident (1:0-1:3) foo
  ident (1:4-1:7) bar
`;

exports[`prop access parses various idents 1`] = `
member . (1:0-1:5)
  ident (1:0-1:3) foo
  ident (1:4-1:5) Ï€
`;

exports[`prop access parses various idents 2`] = `
member . (1:0-1:9)
  ident (1:0-1:3) foo
  ident (1:4-1:9) $spam
`;

exports[`prop access parses various idents 3`] = `
member . (1:0-1:9)
  ident (1:0-1:3) foo
  ident (1:4-1:9) _spam
`;

exports[`prop access parses with taken keywords 1`] = `
member . (1:0-1:9)
  ident (1:0-1:3) foo
  ident (1:4-1:9) match
`;

exports[`prop access parses with taken keywords 2`] = `
member . (1:0-1:6)
  ident (1:0-1:3) foo
  ident (1:4-1:6) fn
`;

exports[`prop access parses with taken keywords 3`] = `
block fn (1:0-1:22)
  assign = (1:3-1:15)
    ident (1:3-1:6) foo
    member . (1:7-1:15)
      ident (1:7-1:10) rec
      ident (1:11-1:15) list
  :
  ident (1:17-1:22) false
`;
